File: 0x00 基本算法/0x07 贪心/AcWing111.py
from heapq import heappush, heappop
N = 50006
def main():
    n = int(input())
    cow, ans, s = [0]*N, [0]*N, [], 
    for i in range(1, n+1):
        l, r = map(int, input().split())
        cow[i] = [i, l, r, 0]
    cow[1:n+1] = sorted(cow[1:n+1], key=lambda x: x[1])
    for i in range(1, n+1):
        num = len(s)
        if num and s[0][0]<cow[i][1]:
            cow[i][3] = s[0][1]
            heappop(s)
            heappush(s, [cow[i][2], cow[i][3]])
            continue
        num += 1; cow[i][3] = num
        heappush(s, [cow[i][2], num])
    print(len(s))
    for i in range(1, n+1): ans[cow[i][0]] = cow[i][3]
    for i in range(1, n+1): print(ans[i])
    
main()

File: 0x00 基本算法/0x07 贪心/AcWing111.cpp
#include <iostream>
#include <queue>
#include <algorithm>
using namespace std;
const int N = 50006;
struct COW {
  int id, l, r, ans;
  bool operator < (const COW x) const {
    return l < x.l;
  }
} cow[N];
priority_queue<pair<int, int> > s; // 默认为大根堆

int main() {
  int n, ans[N];
  cin >> n;
  for (int i=1; i<=n; i++) {
    cow[i].id = i;
    cin >> cow[i].l >> cow[i].r;
  }
  sort(cow+1, cow+n+1);
  // 扫描s，找到一畜栏：满足当前牛开始吃草时间>=畜栏中最后一头牛结束吃草的时间
  for (int i=1; i<=n; i++) {
    int num = s.size();
    if (num && -s.top().first < cow[i].l) {
      cow[i].ans = s.top().second;
      s.pop();
      s.push(make_pair(-cow[i].r, cow[i].ans));
      continue;
    }
    cow[i].ans = ++num;
    s.push(make_pair(-cow[i].r, num));
  }
  cout << s.size() << endl;
  for (int i=1; i<=n; i++) ans[cow[i].id] = cow[i].ans;
  for (int i=1; i<=n; i++) cout << ans[i] << endl;
  return 0;
}